presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineSemiJoinDistributionType.java

```diff
@@ -84,13 +128,59 @@ public PlanNode visitDelete(DeleteNode node, RewriteContext<Void> context)
+            possibleJoinNodes.add(getSemiJoinNodeWithCost(node.withDistributionType(REPLICATED)));
+            possibleJoinNodes.add(getSemiJoinNodeWithCost(node.withDistributionType(PARTITIONED)));
 
-            return SemiJoinNode.DistributionType.REPLICATED;
+            // Using Ordering to facilitate rule determinism
+            Ordering<PlanNodeWithCost> planNodeOrderings = costComparator.forSession(session).onResultOf(PlanNodeWithCost::getCost);
+            return planNodeOrderings.min(possibleJoinNodes).getPlanNode();
+        }
+
+        private PlanNodeWithCost getSemiJoinNodeWithCost(SemiJoinNode possibleJoinNode)
+        {
+            TypeProvider types = symbolAllocator.getTypes();
+            StatsProvider stats = statsProvider;
+            boolean replicated = possibleJoinNode.getDistributionType().get().equals(REPLICATED);
+            /*
+             *   HACK!
+             *
+             *   Currently cost model always has to compute the total cost of an operation.
+             *   For JOIN the total cost consist of 4 parts:
+             *     - Cost of exchanges that have to be introduced to execute a JOIN
+             *     - Cost of building a hash table
+             *     - Cost of probing a hash table
+             *     - Cost of building an output for matched rows
+             *
+             *   When output size for a JOIN cannot be estimated the cost model returns
+             *   UNKNOWN cost for the join.
+             *
+             *   However assuming the cost of JOIN output is always the same, we can still make
+             *   cost based decisions based on the input cost for different types of JOINs.
+             *
+             *   Although the side flipping can be made purely based on stats (smaller side
+             *   always goes to the right), determining JOIN type is not that simple. As when
+             *   choosing REPLICATED over REPARTITIONED join the cost of exchanging and building
+             *   the hash table scales with the number of nodes where the build side is replicated.
+             */
+            int estimatedSourceDistributedTaskCount = taskCountEstimator.estimateSourceDistributedTaskCount();
+            PlanNodeCostEstimate exchangesCost = calculateJoinExchangeCost(
+                    possibleJoinNode.getSource(),
+                    possibleJoinNode.getFilteringSource(),
+                    stats,
+                    types,
+                    replicated,
+                    estimatedSourceDistributedTaskCount);
+            PlanNodeCostEstimate inputCost = calculateJoinInputCost(
+                    possibleJoinNode.getSource(),
+                    possibleJoinNode.getFilteringSource(),
+                    stats,
+                    types,
+                    replicated,
+                    estimatedSourceDistributedTaskCount);
+            return new PlanNodeWithCost(exchangesCost.add(inputCost), possibleJoinNode);
```

Comment:
Cost calculation belongs to a cost calculator, not to a specific optimizer.

Reply:
This is basically the same as the equivalent rule for join.  But I can extract the exchange + input cost addition to as ingle method cost calculator.